core/embed/rust/src/ui/model_tr/component/share_words.rs

```diff
@@ -0,0 +1,251 @@
+            Paginate,
+        },
+        display::Font,
+        geometry::{Alignment, Offset, Rect},
+    },
+};
+
+use heapless::{String, Vec};
+
+use super::{common::display, scrollbar::SCROLLBAR_SPACE, theme, title::Title, ScrollBar};
+
+const WORDS_PER_PAGE: usize = 3;
+const EXTRA_LINE_HEIGHT: i16 = 2;
+const NUMBER_X_OFFSET: i16 = 5;
+const NUMBER_WORD_OFFSET: i16 = 20;
+const NUMBER_FONT: Font = Font::DEMIBOLD;
+const WORD_FONT: Font = Font::NORMAL;
+const INFO_TOP_OFFSET: i16 = 15;
+const MAX_WORDS: usize = 33; // super-shamir has 33 words, all other have less
+
+/// Showing the given share words.
+pub struct ShareWords<T>
+where
+    T: StringType,
+{
+    area: Rect,
+    title: Child<Title<T>>,
+    scrollbar: Child<ScrollBar>,
+    share_words: Vec<T, MAX_WORDS>,
+    page_index: usize,
+}
+
+impl<T> ShareWords<T>
+where
+    T: StringType,
+{
+    pub fn new(title: T, share_words: Vec<T, MAX_WORDS>) -> Self {
+        let mut instance = Self {
+            area: Rect::zero(),
+            title: Child::new(Title::new(title)),
+            scrollbar: Child::new(ScrollBar::to_be_filled_later()),
+            share_words,
+            page_index: 0,
+        };
+        let page_count = instance.total_page_count();
+        let scrollbar = ScrollBar::new(page_count);
+        instance.scrollbar = Child::new(scrollbar);
+        instance
+    }
+
+    fn word_index(&self) -> usize {
+        (self.page_index - 2) * WORDS_PER_PAGE
+    }
+
+    fn is_entry_page(&self) -> bool {
+        self.page_index == 0
+    }
+
+    fn is_second_page(&self) -> bool {
+        self.page_index == 1
+    }
+
+    fn is_final_page(&self) -> bool {
+        self.page_index == self.total_page_count() - 1
+    }
+
+    fn total_page_count(&self) -> usize {
+        let word_screens = if self.share_words.len() % WORDS_PER_PAGE == 0 {
+            self.share_words.len() / WORDS_PER_PAGE
+        } else {
+            self.share_words.len() / WORDS_PER_PAGE + 1
+        };
+        // Two pages before the words, one after it
+        2 + word_screens + 1
+    }
+
+    fn get_first_text(&self) -> String<100> {
+        build_string!(
+            100,
+            "Write all ",
+            inttostr!(self.share_words.len() as u8),
+            " words in order on recovery seed card."
+        )
+    }
+
+    /// Display the first page with user information.
+    fn paint_entry_page(&mut self) {
+        text_multiline_split_words(
+            self.area.split_top(INFO_TOP_OFFSET).1,
+            &self.get_first_text(),
+            Font::BOLD,
+            theme::FG,
+            theme::BG,
+            Alignment::Start,
+        );
+    }
+
+    fn get_second_text(&self) -> String<100> {
+        build_string!(100, "Do NOT make digital copies!")
+    }
+
+    /// Display the second page with user information.
+    fn paint_second_page(&mut self) {
+        text_multiline_split_words(
```

Comment:
using a `Label` here seems more appropriate, as the area is the same in all cases.

Reply:
Trying right now and it seems `Label` is only for one line - did I miss something?

![image](https://github.com/trezor/trezor-firmware/assets/42543243/26c01250-b390-45c2-9ae1-11791523f3a4)

Reply:
Found out the problem - the height of the `Label` is being determined in `place`, and since we do not know yet what will be there, I had just `"".into()` there, so it allocated just one line.

To me, that is problematic, because I would need to find the longest-possible text I will give to it and place it with this text at the very beginning. This means that `Label` is not a very convenient way of dynamically changing text on the screen.

Therefore, I am still in favor of using `text_multiline_split_words`, which can be called with arbitrary text without limitation

Reply:
Refactored a little bit in c8ff4ed4965a02f0094b120f2057d0220cf90059 - so it uses one common function with supplying `text` and `font` - so it is apparent they all use the same `area`